nixos/modules/virtualisation/incus.nix

```
@@ -0,0 +1,233 @@
+                config = {
+                  source = "/var/lib/incus/storage-pools/default";
+                };
+              }
+            ];
+          }
+        '';
+      };
+
+      socketActivation = lib.mkEnableOption (
+        lib.mdDoc ''
+          socket-activation for starting incus.service. Enabling this option
+          will stop incus.service from starting automatically on boot.
+        ''
+      );
+
+      startTimeout = lib.mkOption {
+        type = lib.types.ints.unsigned;
+        default = 600;
+        apply = toString;
+        description = lib.mdDoc ''
+          Time to wait (in seconds) for incusd to become ready to process requests.
+          If incusd does not reply within the configured time, incus.service will be
+          considered failed and systemd will attempt to restart it.
+        '';
+      };
+    };
+  };
+
+  ###### implementation
+  config = lib.mkIf cfg.enable {
+    # https://github.com/lxc/incus/blob/f145309929f849b9951658ad2ba3b8f10cbe69d1/doc/reference/server_settings.md
+    boot.kernel.sysctl = {
+      "fs.aio-max-nr" = 524288;
+      "fs.inotify.max_queued_events" = 1048576;
+      "fs.inotify.max_user_instances" = 1048576;
+      "fs.inotify.max_user_watches" = 1048576;
+      "kernel.dmesg_restrict" = 1;
+      "kernel.keys.maxbytes" = 2000000;
+      "kernel.keys.maxkeys" = 2000;
+      "net.core.bpf_jit_limit" = 1000000000;
+      "net.ipv4.neigh.default.gc_thresh3" = 8192;
+      "net.ipv6.neigh.default.gc_thresh3" = 8192;
+      "vm.max_map_count" = 262144;
+    };
+
+    boot.kernelModules = [
+      "veth"
+      "xt_comment"
+      "xt_CHECKSUM"
+      "xt_MASQUERADE"
+      "vhost_vsock"
+    ] ++ lib.optionals (!config.networking.nftables.enable) [ "iptable_mangle" ];
+
+    environment.systemPackages = [ cfg.package ];
+
+    # Note: the following options are also declared in virtualisation.lxc, but
+    # the latter can't be simply enabled to reuse the formers, because it
+    # does a bunch of unrelated things.
+    systemd.tmpfiles.rules = [ "d /var/lib/lxc/rootfs 0755 root root -" ];
+
+    security.apparmor = {
+      packages = [ cfg.lxcPackage ];
+      policies = {
+        "bin.lxc-start".profile = ''
+          include ${cfg.lxcPackage}/etc/apparmor.d/usr.bin.lxc-start
+        '';
+        "lxc-containers".profile = ''
+          include ${cfg.lxcPackage}/etc/apparmor.d/lxc-containers
+        '';
+      };
+    };
+
+    systemd.services.incus = {
+      description = "Incus Container and Virtual Machine Management Daemon";
+
+      wantedBy = lib.optional (!cfg.socketActivation) "multi-user.target";
+      after = [
+        "network-online.target"
+        "lxcfs.service"
+      ] ++ (lib.optional cfg.socketActivation "incus.socket");
+      requires = [
+        "network-online.target"
+        "lxcfs.service"
+      ] ++ (lib.optional cfg.socketActivation "incus.socket");
+
+      path = lib.optional config.boot.zfs.enabled config.boot.zfs.package;
```

Semantically speaking, it's not like we're planning to set these to the empty string if the predicates are false, it's more like we don't want to set them at all.

```suggestion
      wantedBy = lib.mkIf (!cfg.socketActivation) "multi-user.target";
      after = [
        "network-online.target"
        "lxcfs.service"
      ] ++ (lib.optional cfg.socketActivation "incus.socket");
      requires = [
        "network-online.target"
        "lxcfs.service"
      ] ++ (lib.optional cfg.socketActivation "incus.socket");

      path = lib.mkIf config.boot.zfs.enabled config.boot.zfs.package;
```